Conform pouchdb find operators to couchdb's

AlansCodeLog opened the pull request:
This pull request:

- Fixes pouchdb to conform to [couchdb's operator limitations](https://docs.couchdb.org/en/stable/api/database/find.html#condition-operators) for how it matches against a value, specifically `$size`, `$regex`, and `$mod`. I looked over the other operators and they seem to be fine. 
- Adds a function to validate the selector and check all operator value types (e.g. `$exists` can only be queried by a boolean, `$and` can only be an array, etc.) *before* running / sending the query. 

Notes:

I added a new file `validateSelector` to `pouchdb-find` since there did not seem to be a better place to put it and moved `$mod`'s checks there but did not move the check to see if a "matcher" exists (i.e. the operator is supported by pouch). I left that where it was in `in-memory-filter` because it should be possible to query couchdb with operators that pouchdb does not yet support.

`$regex` is of course also allowed to be an instance of `RegExp` ~~so the validation function won't save users from querying couchdb with one~~ (I fixed it), but I have an idea for how to allow cross-compatible queries. Will file a separate issue.

Other:

The .gitignore is not ignoring all temp db folders generated if they happen to still be around because I stopped a test, etc. Really annoying. I changed mine locally to `tmp*` instead of `tmp`. I can add the changes to this PR or would you prefer a separate one? Or was there some reason for it being just `tmp`?

Also being a double quotes person I tried as much as possible to use single quotes like most of the rest of the source code (though tests has quite the mix), but eslint is not set up with any quote style so some might have slipped through.

daleharvey commented:
Just some quick comments, not much on the code itself, will look at that shortly

    The .gitignore is not ignoring all temp db folders generated if they happen to still be around because I stopped a test, etc. Really annoying. I changed mine locally to tmp* instead of tmp.

Yeh has annoyed me, no reason, feel free to include in this PR

    Also being a double quotes person I tried as much as possible to use single quotes like most of the rest of the source code

Single Quotes was a mistake, feel free to use double and can move towards eslint enforcing that at some point

AlansCodeLog commented:
Cool, I'll add the .gitignore commit then.  

One other thing, I just realized, the validation can be made to error if the user tries to send a regex over http. I can just add a parameter to tell it whether it's in "http" or "local" mode. Will edit the commit.

Regarding the single quotes I'll leave them as is for now if you don't mind (it's quite time consuming changing them manually while not touching other code). Also, can more modern stuff, like template strings, be used in the code? Eslint is setup for it, but erred on the side of not using them because I found none, but it would be nice to know if ES6 is/isn't supported in the code. If it is, a few eslint rules that have auto fix would go a long way towards modernizing everything a bit and keeping the style consisten. I can open a separate issue to discuss.

AlansCodeLog commented:
I added an extra check and tests for it so that the validation doesn't recurse into validated operators that are arrays. But in doing so I found other operators that have a problem. Any operator that is using any `arrayContains*` functions are working incorrectly because they are checking values like this ` docFieldValue.indexOf(val) > -1`, when if an object is passed it should be deep checked. 

http tests should pass in the build but local tests will error.

I think the values can be checked with `collate` instead? I'll take a look at this as soon as I can.

AlansCodeLog commented:
Okay I fixed the `arrayContains` issue. Also stumbled into two other issues I fixed as well.

The first is that implicit $eq against objects was not working (#7446, possibly others about deep nested field access, it's hard to search with everything closed). The way couchdb handles this is that it walks through the entire object *until* it meets an operator, whereas pouch was expecting the first key of an object to be an operator. I added test for both implicit and explicit $eq (which was already working) to make it clear they should behave differently with regards to operators nested inside.

The second is #6664 / #8011 / #6879 / possibly others. I fixed it to correctly massage `$or` instead of treating it just like any other `$and`. This allows pouchdb to correctly handle nested `$or`. The resulting selector looks quite different than couchdb when using `explain`, but I'm assuming the $and / $or selectors are massaged like this for a reason since they don't appear in the in memory filters.

Lastly there was a skipped test that I happened to notice didn't make much sense (probably from copying the `$lte` test above it). I fixed it.

daleharvey commented:
So this was very large and intimidating, apologies for taking so long to look at it, I would really like to get this merged, looks like there was only one failing test and have done a PR @ https://github.com/pouchdb/pouchdb/pull/8260 just to see whats up with that against master, it may just be a phantom failure

AlansCodeLog commented:
😅 No problem. It kind of got out of my hands, one thing led to another. My only real worry is the OR selector massaging. With normal sized queries should not be a problem, but am not sure that is the best way to handle it if it wasn't done like that for a reason (would like to know if it was), but it does fix the issue according to the original code's intentions and I would not know enough to fix it some other way (like sticking just that selector to be in memory).

Was there a failing test, like a specific test I mean? I only saw that one failed because it didn't meet the 100% coverage set in the params/command of the test. I assumed that was done on purpose given it was setup that way.

daleharvey commented:
! So I managed to get a little time to test the coverage error was from https://github.com/pouchdb/pouchdb/commit/fe7ffa9cf7306baf74856b74cf7734ebafac25d0#diff-79342eb93496efe970ba307fd40631f3c6c6c8aaa4db5abdd8593ff8ef1a3e16R135 and given how long this has been waiting I figured I would merge first and figure out whats up second since I didnt want this bitrotten, long overdue but thanks so much :)

AlansCodeLog commented:
Forgot about this tbh 😅 . I've actually been working on a boolean expression query lang / parser that can just be run against any object, and thinking back to this code, I would re-iterate that the way it's "massaging" the selectors is dangerous in a worse case scenario (I don't think developers will run into it if they're in control of the queries, but they might if they aren't).

I glanced at the code real quick again now, I'm still not sure why it's done this way. The queries end up mostly being done in-memory anyways. So unless massaging has some purpose I didn't catch, I don't think the small optimizations it does have much of an effect on performance, in fact, past a certain point it might be the opposite. Also although I've not gone to read CouchDB's code, I'm sure it doesn't do this because it doesn't change the query or choke on extremely complex ones.

When my library is done and I have some time, I'll do some perf tests.
